### ISSUE TITLE
Can't create a generic mapper

### ISSUE THREAD

**Issue by JackPaciox** [mapstruct/mapstruct#3187]
### Use case

the generic mapper could be use to give an extra level of abstraction to the project, this will also allow to implement in sprig boot like application a useful abstraction layer on the services side, so programmers could insert a class that suits for most generics operations like CRUD without writing every single time other classes.
This style of programming is very common with the developers that choose to use a Convert instead of a mapper. 
The reason why I am writing this is because I'd like to optimize my code using the mapper strong point of using a single Instance and implement a better Inversion of control inside my code.  

### Generated Code
```java
import org.springframework.stereotype.Component;

import java.util.List;


@Component
public class mapperAdvanceClass<DTO,ENTITY> implements ListMapper<DTO,ENTITY>{

    @Override
    public List<ENTITY> toEntityList(List<DTO> dtoList) {
        return (List<ENTITY>) dtoList;
    }

    @Override
    public List<DTO> toDTOList(List<ENTITY> entities) {
        return (List<DTO>) entities;
    }

    @Override
    public ENTITY toEntity(DTO dto) {
        return (ENTITY) dto;
    }

    @Override
    public DTO toDto(ENTITY entity) {
        return (DTO) entity;
    }
}
```

### Possible workarounds

create a non map class that is generics and implements an Interface, this interface is also implemented on the various mapper that my applications use.

### MapStruct Version

1.5.3.Final

**Comment by Zegveld** [comment]
I don't really understand how this would work.

If I look at the code you want, it is not safe for usage. (see comments below)
```
@Component
public class mapperAdvanceClass<DTO,ENTITY> implements ListMapper<DTO,ENTITY>{

    @Override
    public List<ENTITY> toEntityList(List<DTO> dtoList) {
        return (List<ENTITY>) dtoList; // high chance of a ClassCastException
    }

    @Override
    public List<DTO> toDTOList(List<ENTITY> entities) {
        return (List<DTO>) entities; // high chance of a ClassCastException
    }

    @Override
    public ENTITY toEntity(DTO dto) {
        return (ENTITY) dto; // high chance of a ClassCastException
    }

    @Override
    public DTO toDto(ENTITY entity) {
        return (DTO) entity; // high chance of a ClassCastException
    }
}
``` 

Mapstruct works compile time, so it needs to know exactly with what objects it is dealing to generate the required mapping code.
You can always yourself create an interface that is not annotated by `@Mapper` which would then be the class you refer to from within Spring for dependency injection. Only the interface which knows what objects to map needs to be annotated with `@Mapper`.

For example:
```java
interface Generic<DTO, ENTITY> {
  ENTITY toEntity(DTO dto);
  DTO toDto(ENTITY entity);
}
@Mapper
interface MyMapper extends Generic<MyDto, MyEntity>{}
```
in the above situation a `MyMapperImpl` class is generated with 2 methods: `MyEntity toEntity(MyDto dto)` and `MyDto toDto(MyEntity entity)`.

**Comment by JackPaciox** [comment]
the code I posted about was a Classic converter, for making possible generic conversion between two differnt Enitities of two differnt type. Like DTO and ENTITY. I'd like to see a way of implenting a mapper for different pair of entity using only one implementation, I figured it out that I can achive something very similar using: 

`@Mapper(componentModel = "spring")
interface MyMapper extends Generic<MyDto, MyEntity>
`
this code will allow you to use the mapper inside a service in a generic manner.
